### PR TITLE
3855 move files to s3 bucket

### LINKED ISSUE #3855: Move files to s3 bucket
- [ ] Move files to s3 bucket

_Originally posted by @minotogna in https://github.com/openforis/fra-platform/issues/3374#issuecomment-2219693750_

### PR DESCRIPTION
**Pull request by sorja**
implementation:
- create a bucket to s3
- implement new service `FileStorage` to write and read files
- remove column `file` from `public.file`

todo:
- [ ] src/server/repository/public/ddl/getCreatePublicSchemaDDL.ts update `public.file` with public migration
- investigate if we want to remove `public.file`

todo/check:
- [x] upload and view avatar
- [x] upload and view file from Links and repository
- [x] upload and view file from EditorWYSIWYG

### CODE REVIEW

File: .env.template [hunk 1]
```diff
@@ -52,3 +52,9 @@ SEPAL_PASSWORD="password"
 GEE_PRIVATE_KEY='private key json content'
 
 FRA_MAIL_ENABLED=false
+
+# File S3
+AWS_ACCESS_KEY_ID=your-access-key-id
+AWS_SECRET_ACCESS_KEY=your-secret-access-key
+AWS_REGION=eu-west-1
+S3_BUCKET_NAME=fra-platform-s3
```
Comment (sorja) [inline .env.template:60]: It is possible to test this branch setting these env vars in `.env`

File: src/server/repository/public/file/fields.ts [hunk 2]
```diff
@@ -1,2 +1 @@
-export const fieldsFileSummary = ['id', 'name', 'uuid', 'created_at', 'length(file) as size']
-export const fieldsFile = [...fieldsFileSummary, 'file']
+export const fieldsFileSummary = ['id', 'name', 'uuid', 'created_at' /* 'length(file) as size' */]
```
Comment (sorja) [inline src/server/repository/public/file/fields.ts]: todo get file size, only used when inspecting the file in UI

File: src/test/migrations/steps/20240829112027-step-files-to-s3.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+export default async (client: BaseProtocol) => {
+  try {
+    // This should be changed to all files,
+    // left for testing purpose (all files 5.7gb, this subset around 80mb)
+    const files = await client.query(`
+          select * from public.file
+          where
+          octet_length(file) / 1024.0 / 1024.0 < 2
+          and name ilike '%fin%';
+    `)
```
Comment (sorja) [inline src/test/migrations/steps/20240829112027-step-files-to-s3.ts:60]: note: only subset is migrated for testing purposes

File: src/test/migrations/steps/20240829112027-step-files-to-s3.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
+import * as path from 'path'
```
Comment (sorja) [inline src/test/migrations/steps/20240829112027-step-files-to-s3.ts:1]: major change 1

File: src/server/service/fileStorage/fileStorage.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
+import { Readable } from 'stream'
```
Comment (sorja) [inline src/server/service/fileStorage/fileStorage.ts:1]: major change 2

**Comment by sorja** [comment]
@minotogna 

Purpose of PR:
Remove column `file` from `public.file` and move it to `s3`
This is done after migrating all files to `s3` in migration step:
https://github.com/openforis/fra-platform/pull/3898/files#diff-83be78115ed816ab6b8367dbf7871927d63af070b0ab471dda8ed408d6f24952

We introduce new service `FileStorage`
https://github.com/openforis/fra-platform/pull/3898/files#diff-ae9344079271ecc6b3715c44f69829251ffc276f58f15662569b1b68634d0e8e

With following methods:
`getFile`
`uploadFile`
`getFileSize`
`fileExists`

The base props taken are `{ path?: string, key: string }`
Where the key is the file name in S3, which for us is the UUID of `public.file`
The is optional string that defaults to `public`, later we might have `static` for static files, `tad` etc

**Comment by sorja** [comment]
We also deprecate static file server:
`src/server/service/file_deprecated/index.ts`

This is to migrate the files later to S3

File: src/server/api/cycleData/print/report.ts [hunk 6]
```diff
@@ -75,7 +75,14 @@ const getPdf = async (req: Request, fileName: string): Promise<Buffer> => {
     return pdfBuffer
   }
 
-  return cachedPdfInfo.file.file
+  // Convert Readable to Buffer
+  const chunks = []
+  // eslint-disable-next-line no-restricted-syntax
+  for await (const chunk of cachedPdfInfo.file.file) {
+    chunks.push(chunk)
+  }
```
Comment (minotogna) [inline src/server/api/cycleData/print/report.ts:83]: should we use Promises.each ?

File: src/server/api/file/get.ts [hunk 7]
```diff
@@ -1,6 +1,6 @@
 import { Request, Response } from 'express'
 
-import { FileRepository, fileTypes } from 'server/service/file'
+import { FileRepository, fileTypes } from 'server/service/file_deprecated'
```
Comment (minotogna) [inline src/server/api/file/get.ts:3]: why do we have file_deprecated ?

**Pull request closed without merging**